Dependency changes

SecondSkoll opened the pull request:
Removes canonical-sphinx as a dependency

Replaces with:
* canonical-sphinx-config - a new (very professional) extension that just emulates the configuration setting part of canonical-sphinx without any of the assets and templating. Needs work but it does the trick for the moment - and should cover most of the configuration options currently set in canonical-sphinx and provides defaults for others (might be missing a couple, but works on the current Ulwazi version).
* sphinx-modern-pdf-style - a slightly improved implementation of our PDF settings and configuration, that I wanted to open source for a while. It does some default configuration specifically for PDF output, and offers overrides for projects (like canonical-sphinx-config which will end up setting Canonical branding). Slight regressing, as it's unbranded, but I can update that soon.

Also:
* Rebases #16 and refactors #18
* and makes some config changes to avoid playing with the static path or `html_js/css_files`.

SecondSkoll committed:
feat: removes canonical-sphinx from doc dependencies

SecondSkoll committed:
fix: relative path to updated local theme

SecondSkoll committed:
chore: fix install

SecondSkoll committed:
fix: requirements

SecondSkoll committed:
feat: css compilation (#21)

* feat: build package locally and remotely for preview

* feat: directly add js files

* feat: add make target to compile css from source

* fix: preview and local use with installed package rather than local

* chore: remove manifest - no longer required

* feat: alter initial css files

* chore: change default stylesheet

* chore: remove pink H1

* fix: re-add manifest

* fix: typo

* fix: typo

* chore: rebases #16 (#22)

SecondSkoll commented on Makefile:
Changes to Makefile adds manual CSS compilation target (`make vanilla-main`. Also adds to install sequence to build and install the current ulwazi wheel.

SecondSkoll commented on docs/conf.py:
Changes to `conf.py` implement dependency changes and adds config for myst parser and PDF configuration extension.

SecondSkoll commented on requirements.txt:
Requirements changes removes canonical-sphinx.

SecondSkoll commented on ulwazi/__init__.py:
```diff
@@ -111,6 +113,75 @@ def apply_heading_classes(body_html: str) -> str:
 
     return str(soup)
 
+def apply_admonition_classes(body_html:str) -> str:
```
Rebases admonition styling from #16

SecondSkoll commented on ulwazi/theme/ulwazi/assets/main.scss:
Moves SCSS file to avoid including it in build

SecondSkoll commented on ulwazi/theme/ulwazi/static/css/vanilla-main.css:
Fixes pink H1 - only an issue because this file is relatively manual right now for some reason

SecondSkoll committed:
chore: move CSS files to stylesheet section in theme.toml

SecondSkoll commented on ulwazi/__init__.py:
```diff
@@ -42,13 +42,6 @@ def config_inited(app, config):  # noqa: ANN401
         except ModuleNotFoundError:  # noqa: PERF203
             print(f"{package} not found.\n{package} will not be configured.")
 
-    extra_css = [
```
Moves CSS from setup to theme.toml

SecondSkoll reviewed (commented):
Adding clarification for changes

SecondSkoll committed:
Merge branch 'main' into secondskoll_dependency-changes

izmalk commented on Makefile:
```diff
@@ -88,8 +88,9 @@ pa11y-install:
 
-build-css:
-	npx sass --load-path=node_modules ulwazi/theme/ulwazi/static/css/main.scss ulwazi/theme/ulwazi/static/css/vanilla-main.css
+install: $(VENVDIR)
```
There is a duplication with a target a few lines below. Can we combine them together?

SecondSkoll replied:
Fixed, it was a result of resolving merge conflicts with the main branch.

SecondSkoll committed:
fix: duplicate make target

SecondSkoll committed:
fix: clean command dep

SecondSkoll committed:
fix: remove additions to static path

SecondSkoll commented on ulwazi/__init__.py:
```diff
@@ -12,10 +12,6 @@
 def setup(app):
     app.add_html_theme('ulwazi', str(THEME_PATH))
 
-    # Register static files path
-    static_path = str(THEME_PATH / "static")
-    app.config.html_static_path.append(static_path)
-
```
By adding css files to the stylesheet config in theme.toml, these files are available for the built documentation, by adding the JS files via `app.add_js_file(<file>)` as below, the same is true of the JS files.

izmalk replied:
I think I've [tried this](https://github.com/canonical/ulwazi/pull/19/files#diff-1f6bd59e1fb6662a4ae915ddd83d72217445f5166d4d7ac845bcdce49fac1c2aR4) locally with the CSS files, and it didn't work.

SecondSkoll replied:
I you have `canonical-sphinx` installed and added as an extension, none of this will work.

SecondSkoll replied:
You can see from the RTD preview that it does work.

Edit: Though I know this is a lot of changes at once and testing pieces is easier to check and understand, there was a reason I originally did this in a specific order - I would make sure `canonical-sphinx` is removed because it really does mess with a lot of things, and makes it hard to see what is happening.

izmalk commented on Makefile:
```diff
@@ -206,6 +202,33 @@ allmetrics: html
+	npm install ;\
+	else \
+	echo "Ensure npm is installed" ;\
+	fi
```
This doesn't work for me. Shows the following error:

```
(.venv) izmalk@Lenovo-Fortress-Lin2:~/test/sphinx/theme/ulwazi$ make npm-install
/usr/bin/npm
/bin/sh: 1: [: -eq: unexpected operator
Ensure npm is installed
(.venv) izmalk@Lenovo-Fortress-Lin2:~/test/sphinx/theme/ulwazi$ make vanilla-main
/usr/bin/npm
/bin/sh: 1: [: -eq: unexpected operator
Ensure npm is installed
echo "Compiling SCSS to CSS..."
Compiling SCSS to CSS...
/bin/sh: 1: [: Illegal number: npm-install
Error: sass not found. Please install sass first.
make: *** [Makefile:217: vanilla-main] Error 1
(.venv) izmalk@Lenovo-Fortress-Lin2:~/test/sphinx/theme/ulwazi$ 
```

I managed to make it work with the following change:

```suggestion
    @command -v npm &> /dev/null 2>&1 || { echo >&2 "Ensure npm is installed"; exit 1;} 
    npm install ;
```

I'm not sure which way is better and why the original doesn't work. Please advise.

SecondSkoll replied:
It looks like it concatenates the make command, so the environment checked by `$?` is the previous shell command used rather than the `@command` line. Good fix, I'll adjust the make vanilla-main command too, as I think that also has issues.

Thanks!

SecondSkoll commented on Makefile:
```diff
@@ -206,6 +202,33 @@ allmetrics: html
+	fi
+
+	@echo "SCSS compilation complete!"
+
```
```suggestion
npm-install:
	@command -v npm >/dev/null 2>&1 || { echo >&2 "Error: 'npm' not found. Please install it."; exit 1; }
	@npm install

vanilla-main: npm-install
	echo "Compiling SCSS to CSS..."

	@echo "Using local sass..."
	@./node_modules/.bin/sass ulwazi/theme/ulwazi/assets/main.scss ulwazi/theme/ulwazi/static/css/vanilla-main.css

	@echo "SCSS compilation complete!"
```

Removes bash `if` control statements due to unexpected issues in conditions.

SecondSkoll replied:
Also uses the `sass` module installed by npm in the project, rather than optionally using the global version.

SecondSkoll committed:
fix: make vanilla-main command and deps

SecondSkoll committed:
chore: change h1 color to validate changes

SecondSkoll committed:
chore: color change

nbellol reviewed (approved):
LGTM! everything works

izmalk commented on ulwazi/theme/ulwazi/static/css/vanilla-main.css:
```diff
@@ -11858,4 +11858,8 @@ ol.u-text-max-width,
 
+h1 {
+  color: yellowgreen;
+}
```
Why do we have this?
Is this needed to make sure we've built the file correctly?
So that if the pre-built version is used, then it has Green h1?
I'm asking because it's actually green for me if I build it locally, even after clean up and `make vanilla-main`

izmalk replied:
Discussed with @nbellol - it seems we need it for testing.
I suggested another approach - to use a unique special style and add a text with this custom style assigned in our sample content. This way we can test the theme being built properly yet have it not affecting real documentation (unless it uses this special style).

izmalk replied:
Fixed in https://github.com/canonical/ulwazi/pull/20/commits/f612d2b7f25ed1f8ab75122657ba4f29b21259e0

izmalk commented on ulwazi/theme/ulwazi/assets/main.scss:
```diff
@@ -0,0 +1,8 @@
+
+h1 {
+    color: yellowgreen;
+}
```
Are you sure we need it in both here and in `ulwazi/theme/ulwazi/static/css/vanilla-main.css`?

izmalk replied:
Fixed in https://github.com/canonical/ulwazi/pull/20/commits/f612d2b7f25ed1f8ab75122657ba4f29b21259e0

izmalk commented on Makefile:
```diff
@@ -56,7 +56,7 @@ venv:
         epub serve clean fclean clean-sp clean-doc spelling spellcheck linkcheck woke \
         allmetrics pa11y pdf-prep-force pdf-prep pdf vale-install vale \
-		update
+		update compile-scss
```
We need to update the `help` too, in the beginning.

izmalk replied:
Fixed in https://github.com/canonical/ulwazi/pull/20/commits/cbec10e9bd0c0fba8a349270b794c0d37db6aa24

nbellol committed:
move propagation test to not the header

izmalk committed:
Update Makefile

Add `npm-install` and `vanilla-main` targets description to the help.
Additionally, minor updates - add `rebuild` description and update `fclean` description.

izmalk reviewed (approved):
LGTM
